pip/pip-338.md

```diff
@@ -0,0 +1,172 @@
+               path = listenerName == null ? path : path + "&listenerName=" + listenerName;
+               redirect = new URI(path);
+           }
+    
+```
+
+## Need to have lookupListener
+The Pulsar broker has the concept of advertised listeners representing broker endpoints that are discoverable by, and accessible to, Pulsar clients. In addition, one listener may be designated as the internal listener, to be used for broker-to-broker communication. Listeners may be understood as named routes to the broker. If a listener name is not specified, the response contains the endpoint for the internal listener, or the first listener for the protocol.
+
+But the challenge arises when it considers the same internalListener for lookup requests redirects. This may result in returning an unresolvable broker service address to the clients.
+It may not be possible to pass the listenerName from the clients, which consequently may lead to cluster downtime.
+
+it might not be feasible to have the listerner config at the client side in every tech stack or connector.
+Also,  Currently, there is no option to pass the listener while using the admin APIs. As admin APIs can be called from an external network as well, the use of the internal listener’s broker service URL can lead all admin operations to get affected.
+
+Moreover, as per current code provided above, even if the client provides a listenerName/listener header, the redirect urls are taken from the internal listener or the first listener with http protocol which may not be a client-resolvable http address.
+
+# Proposed Solution
+To fix inconsistency with all protocol’s service url, we need to include all broker addresses associated with the listener name while returning the lookupResult, rather than solely the service URL
+
+Also, Introduce a broker-level property to determine the appropriate listener for lookup calls if listenerName is missing from the client request. This can help deal with the side effects of not having listerName on the client side when multiple advertisedListeners are being used.
+
+Also, this can help to determine the broker service URL that needs to be returned in case of lookup redirects and also for using Pulsar admin API calls from outside the network.
+
+This will help in having a more transparent listener identifier for internal and external communications.
+
+
+# Design & Implementation Details
+## Approach
+**1. Fix Code Flow - Inconsistency with listener usage with protocols other than pulsar and pulsar+ssl** 
+
+Currently, lookup result return the broker service url corresponding to the internal listerner in case of http or https protocol urls. To address the issue in the workflow, it is necessary to include all broker addresses associated with the listener name, rather than solely the service URL. There can be a check if all broker addresses corresponding to the given listener is null, then use the default listener(discussed below)
```

Comment:
For topic lookups a Pulsar binary protocol endpoint is always returned. It doesn't make sense to return all endpoints for topic lookups.

Reply:
Agreed, its confusing to see the http urls present in lookup response, and we are lacking historic info on why they are present. But _if and when_ they are already present, they should be resolvable based in the ask of the request/default config.

Reply:
Yes you are right, this is the bug that would need fixing, mentioned in the previous comment https://github.com/apache/pulsar/pull/22039#discussion_r1487402439 .